Tools/TestWebKitAPI/Tests/IPC/ArgumentCoderTests.cpp

```diff
@@ -0,0 +1,153 @@
+    }
+
+    void encode(IPC::Encoder& encoder) && {
+        encoder << uint32_t(0);
+        ++m_counterValues.encodingRValue;
+    }
+
+    static std::optional<EncodingCounter> decode(IPC::Decoder&) { return std::nullopt; }
+
+    CounterValues& m_counterValues;
+};
+
+TEST(ArgumentCoder, MoveSemantics)
+{
+    IPC::Encoder encoder(static_cast<IPC::MessageName>(0), 42);
+
+    auto test =
+        [&encoder](unsigned expectedEncodingCount, auto&& createFunctor)
+        {
+            // Encoding as an lvalue.
+            {
+                EncodingCounter::CounterValues counterValues;
+                auto object = createFunctor(counterValues);
+                encoder << object;
+
+                counterValues.assertEqual(expectedEncodingCount, 0);
+            }
+
+            // Encoding as an rvalue moved into the encoding call.
+            {
+                EncodingCounter::CounterValues counterValues;
+                auto object = createFunctor(counterValues);
+                encoder << WTFMove(object);
+
+                counterValues.assertEqual(0, expectedEncodingCount);
+            }
+
+            // Encoding as an rvalue created in the encoding call.
+            {
+                EncodingCounter::CounterValues counterValues;
+                encoder << createFunctor(counterValues);
+
+                counterValues.assertEqual(0, expectedEncodingCount);
+            }
+        };
+
+    test(1,
+        [](auto& counterValues)
+        {
+            return EncodingCounter { counterValues };
+        });
+
+    test(1,
+        [](auto& counterValues)
+        {
+            return std::optional<EncodingCounter> { EncodingCounter { counterValues } };
+        });
+
+    test(1,
+        [](auto& counterValues)
+        {
+            return std::pair<unsigned, EncodingCounter> { 0, EncodingCounter { counterValues } };
+        });
+
+    test(2,
+        [](auto& counterValues)
+        {
+            return std::tuple<EncodingCounter, unsigned, EncodingCounter> {
+                EncodingCounter { counterValues }, 0,
+                EncodingCounter { counterValues },
+            };
+        });
+
+    test(4,
+        [](auto& counterValues)
+        {
+            return std::array<EncodingCounter, 4> {
+                EncodingCounter { counterValues },
+                EncodingCounter { counterValues },
+                EncodingCounter { counterValues },
+                EncodingCounter { counterValues },
+            };
+        });
+
+    test(16,
+        [](auto& counterValues)
+        {
+            Vector<EncodingCounter> counters;
+            for (unsigned i = 0; i < 16; ++i)
+                counters.append(EncodingCounter { counterValues });
+            return counters;
+        });
+
+    test(1,
+        [](auto& counterValues)
+        {
+            return std::variant<EncodingCounter, unsigned> { EncodingCounter { counterValues } };
+        });
+}
+
+} // namespace TestWebKitAPI
```

Comment:
Something that might be useful would be to try if encoding containers of const types still compiles. This isn't invoked much in current webkit, and I'm not 100% sure decoding works, but just if you're looking more to cover.

Reply:
You mean something like `Vector<const T>`? My hunch is reference collapsing would yield a const lvalue reference. I can add tests.